Reland [clang][modules-driver] Add scanner to detect C++20 module presence

naveen-seth opened the pull request:
This patch is part of a series to natively support C++20 module usage from the Clang driver (without requiring an external build system). This introduces a new scanner that detects C++20 module usage in source files without using the preprocessor or lexer.

For now, it is enabled only with the `-fmodules-driver` flag and serves solely diagnostic purposes. In the future, the scanner will be enabled for any (modules-driver compatible) compilation with two or more inputs, and will help the driver determine whether to implicitly enable the modules driver.

Since the scanner adds very little overhead, we are also exploring enabling it for compilations with only a single input. This approach could allow us to detect `import std` usage in a single-file compilation, which would then activate the modules driver. For performance measurements on this, see
https://github.com/naveen-seth/llvm-dev-cxx-modules-check-benchmark.

RFC:
https://discourse.llvm.org/t/rfc-modules-support-simple-c-20-modules-use-from-the-clang-driver-without-a-build-system

This patch relands commit ded1426. The CI failure is resolved by removing the compatibility warning for using the `-fmodules-driver` flag with pre-C++20 standards, which also better aligns its behavior with other features/flags supported only in newer standards.

naveen-seth committed:
Reland [clang][modules-driver] Add scanner to detect C++20 module presence

This patch is part of a series to natively support C++20 module usage
from the Clang driver (without requiring an external build system).
This introduces a new scanner that detects C++20 module usage in
source files without using the preprocessor or lexer.

For now, it is enabled only with the `-fmodules-driver` flag and
serves solely diagnostic purposes. In the future, the scanner will be
enabled for any (modules-driver compatible) compilation with two or
more inputs, and will help the driver determine whether to implicitly
enable the modules driver.

Since the scanner adds very little overhead, we are also exploring
enabling it for compilations with only a single input.
This approach could allow us to detect `import std` usage in a
single-file compilation, which would then activate the modules driver.
For performance measurements on this, see
https://github.com/naveen-seth/llvm-dev-cxx-modules-check-benchmark.

RFC:
https://discourse.llvm.org/t/rfc-modules-support-simple-c-20-modules-use-from-the-clang-driver-without-a-build-system

This patch relands commit ded1426. The CI failure is resolved by
removing the compatibility warning for using the `-fmodules-driver`
flag with pre-C++20 standards, which also better aligns its behavior
with other features/flags supported only in newer standards.

naveen-seth committed:
Fix comment formatting

naveen-seth committed:
Revert changes to regression test

cor3ntin commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -4417,6 +4424,174 @@ void Driver::handleArguments(Compilation &C, DerivedArgList &Args,
+  if (unsigned Len = isLineSplice(Ptr); Len) {
+    Ptr += Len;
+    return true;
+  }
+  return false;
+}
+
+static bool trySkipDirective(const char *&Ptr) {
+  if (*Ptr != '#')
+    return false;
+
+  ++Ptr;
+  skipLine(Ptr);
+  return true;
+}
+
+static bool trySkipLineComment(const char *&Ptr) {
+  if (Ptr[0] != '/' || Ptr[1] != '/')
+    return false;
+
+  Ptr += 2;
+  skipLine(Ptr);
+  return true;
+}
+
+static bool trySkipBlockComment(const char *&Ptr) {
+  if (Ptr[0] != '/' || Ptr[1] != '*')
+    return false;
+
+  Ptr += 2;
+  while (*Ptr != '\0') {
+    if (Ptr[0] == '*' && Ptr[1] == '/') {
+      Ptr += 2; // '*/'
+      return true;
+    }
+    ++Ptr;
+  }
+  return true;
+}
+
```
That code is similar to what we do in DependencyDirectivesScanner.cpp, could we try to remove some redundancies here? @Bigcheese

Bigcheese replied:
Yeah, there's a lot in common here. The difference is these don't take an end and rely on the null terminator due in part to terminating as soon as we can determine if it's a named module source file or not. I think it's fine to use the end versions though as we don't actually dereference any memory past what's currently being looked at.

It's worth looking at what can be reused here. I think it's fine to just expose some of the functions from `DependencyDirectivesScanner.cpp` under the `clang::dependency_directives_scan` namespace. Or even move `scanBufferForCXXModuleUsage` there entirely.

naveen-seth committed:
Merge branch 'llvm:main' into dev-modules-driver-named-module-detection-reland

naveen-seth committed:
Refactor/migrate to DependencyDirectivesScanner.cpp for better code reuse

This moves the check for C++20 modules usage into
DependencyDirectivesScanner.cpp to better reuuse existing code.
This now also leverages the dependency directives scanner to lex the
module text line.

naveen-seth commented:
This now also leverages the `Scanner` in `DependencyDirectivesScanner.cpp` to reuse its logic for properly lexing module related directives. For this to happen, I had to first land two fixes with #148674 and #148685.

naveen-seth committed:
Revert include of "clang/Basic/CharInfo.h" in Driver.cpp

Because `scanInputForCXX20ModuleUsage` has been moved out of
Driver.cpp, including "CharInfo.h" here is not required anymore.
(This also fixes formatting in one comment).
